src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ingroovy/StepInGroovy.java

```diff
@@ -0,0 +1,167 @@
+        /**
+         * Groovy source file compiled with normal Groovy compiler, not CPS.
+         * Useful for introspection.
+         */
+        private final StepInGroovyScript compiled;
+        /**
+         * The 'call' method on {@link #compiled}
+         */
+        private final Method callMethod;
+        private final List<Parameter> params = new ArrayList<>();
+
+        /**
+         * @param fqcn
+         *      Fully qualified class name of Groovy class that defines this step, to be loaded
+         *      from trusted CPS content root.
+         */
+        public StepDescriptorInGroovy(String fqcn) throws IOException {
+            this.className = fqcn;
+            this.functionName = FilenameUtils.getBaseName(fqcn);
+            this.compiled = GroovyCompiler.get().parse(this);
+            this.callMethod = findCallMethod();
+
+            Class<?>[] p = callMethod.getParameterTypes();
+            /*
+                TODO: ClassDescriptor.loadParameterNames requires .class file to be present, which isn't
+                We need AST tranformation to capture CapturedParameterNames
+             */
+//            String[] n = ClassDescriptor.loadParameterNames(callMethod);
+//            assert p.length==n.length;
+//            for (int i=0; i<n.length; i++) {
+//                params.add(new Parameter(p[i],n[i]));
+//            }
+
+            for (int i=0; i<p.length; i++) {
+                params.add(new Parameter(p[i],"var"+i));
+            }
+        }
+
+        @Override
+        public Map<String, Object> singleArgument(Object arg) {
+            if (params.size()==1)
+                return Collections.singletonMap(params.get(0).name, arg);
+            else
+                return null;
+        }
+
+        private Method findCallMethod() {
+            for (Method m : compiled.getClass().getMethods()) {
+                if (m.getName().equals("call"))  {
+                    return m;
+                }
+            }
+            throw new IllegalArgumentException(getClassName()+" does not have a public call method");
+        }
+
+        /**
+         * Parameters defined on this step, which are those that are defined on the 'call' method.
+         */
+        public List<Parameter> getParameters() {
+            return params;
+        }
+
+        /**
+         * {@link #clazz} doesn't uniquely identify a descriptor,
+         * so we need more things to make it unique.
+         */
+        @Override
+        public String getId() {
+            return super.getId()+":"+functionName;
+        }
+
+        /**
+         * If the groovy code takes the closure parameter in the end, that's a sign that we take
+         * body.
+         */
+        @Override
+        public boolean takesImplicitBlockArgument() {
+            return params.isEmpty() && Closure.class.isAssignableFrom(params.get(params.size()-1).type);
+        }
+
+        @Override
+        public Set<Class<?>> getRequiredContext() {
```

Comment:
Would be interesting to see if we can figure out required context from the step itself, but yeah, that's not viable for now.